### PR TITLE
add dialog to Launcher

### LINKED ISSUE #11783: Add a visible notice when user opens more than one JabRef instance
**Is your suggestion for improvement related to a problem? Please describe.**

I wanted to run a portable version of JabRef, but I already had running JabRef in IntelliJ.

When I double clicked on the executable file in portable version, nothing happended. I didn't give up, and started it from terminal. In terminal there was:

```
Messages are not initialized before accessing key: Display help on command line options
2024-09-18 08:10:49 [main] org.jabref.Launcher.handleMultipleAppInstances()
INFO: Arguments passed on to running JabRef instance. Shutting down.
```

I couldn't understood this at the beginning, but then I finally understood that I had another instance of JabRef running.

`Arguments passed on to running JabRef instance` - WTF is this? 1) I didn't pass any arguments, and 2) I was certain that exactly this executable wasn't run before.

**Describe the solution you'd like**

1) Refactor the message when user starts JabRef a second time.
2) Make a dialog with that message

**Additional context**
-

### PR DESCRIPTION
**Pull request by leaf-soba**
- Closes #11783
- when user open two instances of JabRef, second one will open dialog saying Localization.lang("Another JabRef instance is already running. Please switch to that instance.")
- since `DialogService` didn't initialize at that life cycle, I tried some way to get around it but all failed, such as:
1. NPE
```
dialogService.showInformationDialogAndWait(Localization.lang("Opening multiple JabRef instance without arguments"), Localization.lang("Another JabRef instance is already running. Please switch to that instance."));
```
2. `java.lang.IllegalStateException: This operation is permitted on the event thread only; currentThread = main`
```
JabRefDialogService dialogService = new JabRefDialogService(new BaseWindow());
dialogService.showInformationDialogAndWait(Localization.lang("Opening multiple JabRef instance without arguments"), Localization.lang("Another JabRef instance is already running. Please switch to that instance."));
```
3. Unexpected exception: java.lang.IllegalStateException: Toolkit not initialized
```
UiTaskExecutor.runInJavaFXThread(() -> {
    DialogService dialogService = Injector.instantiateModelOrService(DialogService.class);
    dialogService.showErrorDialogAndWait(Localization.lang("Another JabRef instance is already running. Please switch to that instance."));
}));
```
4. Nothing happened; it seems the JavaFXThread wasn't created before the instance was terminated
```
                    FallbackExceptionHandler.installExceptionHandler((exception, thread) -> {
                        UiTaskExecutor.runInJavaFXThread(() -> {
                            DialogService dialogService = Injector.instantiateModelOrService(DialogService.class);
                            dialogService.showErrorDialogAndWait("Another JabRef instance is already running. Please switch to that instance.");
                        });
                    });
```
- so I add `JOptionPane` to open a dialog, but I think it may not the best practice in this situation.

---
edit: I changed to JavaFX now:

---
edit:
1. removed swing based on comment
2. open a dialog in exist instance

---
edit: fix base on devcall
1. fix a little bug in `shutdownThreadPools()` when `taskExecutor` is null, it can't be shutdown.
2. open a dialog after `initialize()`

### Mandatory checks

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/Launcher.java [hunk 1]
```diff
@@ -11,6 +11,9 @@
 import java.util.List;
 import java.util.Map;
 
+import javax.swing.JOptionPane;
+import javax.swing.SwingUtilities;
```
Comment (subhramit) [inline src/main/java/org/jabref/Launcher.java:15]: I don't think we use swing in JabRef. Can you find JavaFX equivalents and use them?

Comment (Siedlerchr) [reply src/main/java/org/jabref/Launcher.java:15]: Yes, but the javafx is not yet initialized only in JabrefGui.

Comment (Siedlerchr) [reply src/main/java/org/jabref/Launcher.java:15]: A workaround would be to add a consumer  lambda as a parameter to the method handleMultipleAppInstances  and then pass that down to JabRefGUI where you can use it to show a dialog
I did a similar thing for the FallbackExceptionHandler

Comment (leaf-soba) [reply src/main/java/org/jabref/Launcher.java:15]: > A workaround would be to add a consumer lambda as a parameter to the method handleMultipleAppInstances and then pass that down to JabRefGUI where you can use it to show a dialog I did a similar thing for the FallbackExceptionHandler

I have tried your solution yesterday I forget to add it into origin post:
4. Nothing happened; it seems the JavaFXThread wasn't created before the instance was terminated
```
                    FallbackExceptionHandler.installExceptionHandler((exception, thread) -> {
                        UiTaskExecutor.runInJavaFXThread(() -> {
                            DialogService dialogService = Injector.instantiateModelOrService(DialogService.class);
                            dialogService.showErrorDialogAndWait("Another JabRef instance is already running. Please switch to that instance.");
                        });
                    });
```

**Comment by calixtus** [comment]
Hi, thanks for your efforts.
There are several constraints to keep in mind:
- Use of Swing is not allowed, except for undo-redo. And we are even thinking of moving away to are more modern framework working with the command design pattern.
- The GUI is constructed in JabRefGui, the Launcher is just there to do the most basic app setup and to call the argument processor, then the GUI.
- there is no proper di framework, only afterburner.fx . Afterburner does not understand a Singleton and cannot just inject without the programmer initializing/running the GUI first and creating the instance for the injection map.

My solution strategy for this would be either to put a message in the ui message list handle it in the GUI and then tear the GUI down again -or - to display the message in the already running instance.

File: src/main/java/org/jabref/Launcher.java (hunk 1 above)
Comment (calixtus) [reply src/main/java/org/jabref/Launcher.java]: Creating the javafx platform is costly and takes time. So displaying a message would require to wait for the platform to be initialized.

**Comment by koppor** [comment]
Adding label devcall since this is more controversial...

**Comment by calixtus** [comment]
DevCall decision:
Hi leaf-soba, we discussed your approach, that looked certainly interesting, but decided against it, because it would be using magic strings to be send to the remote client. We thought about a different approach:
You could check in the Launcher for a second instance and if it exists, add an `uiCommand` to the list sent to JabRefGUI::setup.
This can be intercepted in `JabRefGUI::start` before running `JabRefGUI::initialize` a dialog be shown and the javafx torne down.

We prepared a sketch diff:
```diff
diff --git a/src/main/java/org/jabref/Launcher.java b/src/main/java/org/jabref/Launcher.java
index 1d3cb9cb3e..e5a04684b7 100644
--- a/src/main/java/org/jabref/Launcher.java
+++ b/src/main/java/org/jabref/Launcher.java
@@ -109,6 +109,7 @@ public class Launcher {
                 }

                 List<UiCommand> uiCommands = new ArrayList<>(argumentProcessor.getUiCommands());
+                uiCommands.add(new UiCommand.InstanceAlreadyRunning());
                 JabRefGUI.setup(uiCommands, preferences, fileUpdateMonitor);
                 JabRefGUI.launch(JabRefGUI.class, args);
             } catch (ParseException e) {
diff --git a/src/main/java/org/jabref/gui/JabRefGUI.java b/src/main/java/org/jabref/gui/JabRefGUI.java
index d6a80bf394..1ab30890a1 100644
--- a/src/main/java/org/jabref/gui/JabRefGUI.java
+++ b/src/main/java/org/jabref/gui/JabRefGUI.java
@@ -94,6 +94,12 @@ public class JabRefGUI extends Application {
             });
         });

+        if (uiCommands.stream().anyMatch(UiCommand.InstanceAlreadyRunning.class::isInstance)) {
+            // displayDialog
+            Platform.exit();
+            return;
+        }
+
         initialize();

         JabRefGUI.mainFrame = new JabRefFrame(
diff --git a/src/main/java/org/jabref/logic/UiCommand.java b/src/main/java/org/jabref/logic/UiCommand.java
index 1fc8a699ce..019d09983f 100644
--- a/src/main/java/org/jabref/logic/UiCommand.java
+++ b/src/main/java/org/jabref/logic/UiCommand.java
@@ -12,4 +12,6 @@ public sealed interface UiCommand {
     record OpenDatabases(List<ParserResult> parserResults) implements UiCommand { }

     record AutoSetFileLinks(List<ParserResult> parserResults) implements UiCommand { }
+
+    record InstanceAlreadyRunning() implements UiCommand { }
 }
```

Thanks for your patience!

**Comment by leaf-soba** [comment]
> DevCall decision: Hi leaf-soba, we discussed your approach, that looked certainly interesting, but decided against it, because it would be using magic strings to be send to the remote client. We thought about a different approach: You could check in the Launcher for a second instance and if it exists, add an `uiCommand` to the list sent to `JabRefGUI::setup`. This can be intercepted in `JabRefGUI::start` before running `JabRefGUI::initialize` a dialog be shown and then javafx torne down.
> 
> We prepared a sketch diff:
> 
> ```diff
> diff --git a/src/main/java/org/jabref/Launcher.java b/src/main/java/org/jabref/Launcher.java
> index 1d3cb9cb3e..e5a04684b7 100644
> --- a/src/main/java/org/jabref/Launcher.java
> +++ b/src/main/java/org/jabref/Launcher.java
> @@ -109,6 +109,7 @@ public class Launcher {
>                  }
> 
>                  List<UiCommand> uiCommands = new ArrayList<>(argumentProcessor.getUiCommands());
> +                uiCommands.add(new UiCommand.InstanceAlreadyRunning());
>                  JabRefGUI.setup(uiCommands, preferences, fileUpdateMonitor);
>                  JabRefGUI.launch(JabRefGUI.class, args);
>              } catch (ParseException e) {
> diff --git a/src/main/java/org/jabref/gui/JabRefGUI.java b/src/main/java/org/jabref/gui/JabRefGUI.java
> index d6a80bf394..1ab30890a1 100644
> --- a/src/main/java/org/jabref/gui/JabRefGUI.java
> +++ b/src/main/java/org/jabref/gui/JabRefGUI.java
> @@ -94,6 +94,12 @@ public class JabRefGUI extends Application {
>              });
>          });
> 
> +        if (uiCommands.stream().anyMatch(UiCommand.InstanceAlreadyRunning.class::isInstance)) {
> +            // displayDialog
> +            Platform.exit();
> +            return;
> +        }
> +
>          initialize();
> 
>          JabRefGUI.mainFrame = new JabRefFrame(
> diff --git a/src/main/java/org/jabref/logic/UiCommand.java b/src/main/java/org/jabref/logic/UiCommand.java
> index 1fc8a699ce..019d09983f 100644
> --- a/src/main/java/org/jabref/logic/UiCommand.java
> +++ b/src/main/java/org/jabref/logic/UiCommand.java
> @@ -12,4 +12,6 @@ public sealed interface UiCommand {
>      record OpenDatabases(List<ParserResult> parserResults) implements UiCommand { }
> 
>      record AutoSetFileLinks(List<ParserResult> parserResults) implements UiCommand { }
> +
> +    record InstanceAlreadyRunning() implements UiCommand { }
>  }
> ```
> 
> Thanks for your patience!

I tried your solution, got an error said `ERROR: Unable to close AI service: java.lang.NullPointerException: Cannot invoke "org.jabref.logic.ai.AiService.close()" because "org.jabref.gui.JabRefGUI.aiService" is null`
at
``` 
Platform.exit()
```
It seems this solution add more unnecessary difficulty to this simple issue.If we don't like "magic strings", I think change the string to a formal command string would be better?

**Comment by calixtus** [comment]
Can you push your changes please?
And @InAnYan can you take a look? Seems like the aiService is not shutting down properly at this stage...

**Comment by InAnYan** [comment]
Thanks, Carl, for pointing this out!

@leaf-soba, can you please add `aiService.close()` before the exit?

In theory, it shouldn't be necessary, but AI does a lot of stuff, so better to close it properly

**Comment by InAnYan** [comment]
@leaf-soba, if you have some time, could you post the full stacktrace here? I thought this `aiService.close()` isn't called...

In any ways, quick win: before calling `aiService.close()`, check if it's not null. Something like this:

```java
if (aiService != null) {
	aiService.close();
}
```

**Comment by leaf-soba** [comment]
> Can you push your changes please? And @InAnYan can you take a look? Seems like the aiService is not shutting down properly at this stage...

done, sry I thought I should not push a test code to a branch in this project, so I just described my error which might be confusing.

**Comment by koppor** [comment]
I never thought it would be that much effort to show a simple JavaFX window... Sorry for that!

Do we really need to fire up the main window ( `openWindow()`)? -- Isn't it possible to just show the dialog? "dialog" interpreted in a more broader sence: A small window showing something... -- I would be fine in having another class `extends Application` showing only that. But @Siedlerchr was agagainst it?!

**Comment by Siedlerchr** [comment]
Extending Application here would probably cause more problems than it solves

**Comment by calixtus** [comment]
Extending application (in a separate class) is just the same effort as running JabRefGui. Most computing time is used by creating the ui thread by javafx, so there would be absolutely no benefit.

**Comment by calixtus** [comment]
Only adding complexity that needs to be maintained and a very weird two-javafx-application-setup in one app.

**Comment by leaf-soba** [comment]
> I never thought it would be that much effort to show a simple JavaFX window... Sorry for that!
> 
> Do we really need to fire up the main window ( `openWindow()`)? -- Isn't it possible to just show the dialog? "dialog" interpreted in a more broader sence: A small window showing something... -- I would be fine in having another class `extends Application` showing only that. But @Siedlerchr was agagainst it?!

If I can use swing like second solution it is simple, but we don't like swing.

**Comment by calixtus** [comment]
We decided against swing. Years ago.
In theory everything is simple. Yet it would overturn and ridicule decisions made and ourselves.
Also the Launcher is not part of GUI, but of headless CLI package. If we plan to move forward into multi module build, this would be a major step back to include ui stuff in the Launcher.

**Comment by koppor** [comment]
@leaf-soba Thank you for the efforts invested here. It turned out that the wish causes many code changes. @calixtus, @Siedlerchr and other devleopers in the devcall on Monday invested much time in discussing this. I think, at least two hours in total. Seeing the ongoing discussions and introduced architectural WTFs, it is simply not worth the current and future maintainence effort.

We will come back to the issue if some more users complain.

Internally, we will check our processes to ensure that issues are marked as "good first issues" after they were properly reviewed.

---

Nevertheless, can you modify this PR to NOT do any command passing, but keeping the Java code around the check of `args.lgenth == 0`?

```java
                // There is already a server out there, avoid showing log "Passing arguments" while no arguments are provided.
                if (args.length == 0) {
                    LOGGER.debug("This JabRef instance is already running. Please switch to that instance.");
                } else {
```

**Comment by leaf-soba** [comment]
> @leaf-soba Thank you for the efforts invested here. It turned out that the wish causes many code changes. @calixtus, @Siedlerchr and other devleopers in the devcall on Monday invested much time in discussing this. I think, at least two hours in total. Seeing the ongoing discussions and introduced architectural WTFs, it is simply not worth the current and future maintainence effort.
> 
> We will come back to the issue if some more users complain.
> 
> Internally, we will check our processes to ensure that issues are marked as "good first issues" after they were properly reviewed.
> 
> Nevertheless, can you modify this PR to NOT do any command passing, but keeping the Java code around the check of `args.lgenth == 0`?
> 
> ```java
>                 // There is already a server out there, avoid showing log "Passing arguments" while no arguments are provided.
>                 if (args.length == 0) {
>                     LOGGER.debug("This JabRef instance is already running. Please switch to that instance.");
>                 } else {
> ```

OK, actually I do love doing a lot of changes based on comments and enjoy this times in this issue, since I have learned a lot from it, thank you very much.

**Comment by leaf-soba** [comment]
I think I should keep this code in `org.jabref.gui.JabRefGUI#shutdownThreadPools`?
```
if (taskExecutor != null) {
    taskExecutor.shutdown();
}
```

**Comment by koppor** [comment]
> I think I should keep this code in `org.jabref.gui.JabRefGUI#shutdownThreadPools`?
> 
> ```
> if (taskExecutor != null) {
>     taskExecutor.shutdown();
> }
> ```

I am unsure, but OK for me :)

File: src/main/java/org/jabref/Launcher.java [hunk 2]
```diff
@@ -168,26 +168,29 @@ private static void initLogging(String[] args) {
...
-                // arguments to other instance
-                LOGGER.debug("Passing arguments passed on to running JabRef...");
-                if (remoteClient.sendCommandLineArguments(args)) {
-                    // So we assume it's all taken care of, and quit.
-                    // Output to both to the log and the screen. Therefore, we do not have an additional System.out.println.
-                    LOGGER.info("Arguments passed on to running JabRef instance. Shutting down.");
-                    return false;
+                // There is already a server out there, avoid showing log "Passing arguments" while no arguments are provided.
+                if (args.length == 0) {
+                    LOGGER.debug("This JabRef instance is already running. Please switch to that instance.");
```
Comment (koppor) [inline src/main/java/org/jabref/Launcher.java:180]: Higher log level to ensure that is really output on the console.

```suggestion
                    LOGGER.warn("This JabRef instance is already running. Please switch to that instance.");
```

File: src/main/java/org/jabref/Launcher.java [hunk 3]
```diff
@@ -168,26 +168,29 @@ private static void initLogging(String[] args) {
...
                 LOGGER.debug("Pinging other instance succeeded.");
-                // We are not alone, there is already a server out there, send command line
-                // arguments to other instance
-                LOGGER.debug("Passing arguments passed on to running JabRef...");
-                if (remoteClient.sendCommandLineArguments(args)) {
-                    // So we assume it's all taken care of, and quit.
-                    // Output to both to the log and the screen. Therefore, we do not have an additional System.out.println.
-                    LOGGER.info("Arguments passed on to running JabRef instance. Shutting down.");
-                    return false;
+                // There is already a server out there, avoid showing log "Passing arguments" while no arguments are provided.
```
Comment (koppor) [inline src/main/java/org/jabref/Launcher.java:178]: Can you move that inside the if? (meaning one line down and indent)

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 4]
```diff
@@ -93,7 +93,6 @@ public void start(Stage stage) {
                 dialogService.showErrorDialogAndWait("Uncaught exception occurred in " + thread, exception);
             });
         });
-
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:96]: I think, this is not consistent with the other style in this method, where „everything“ is separated with an empty line. Thus, can you revert this please? 😅

File: src/main/java/org/jabref/Launcher.java [hunk 5]
```diff
@@ -168,26 +168,29 @@ private static void initLogging(String[] args) {
...
+                    LOGGER.debug("Passing arguments passed on to running JabRef...");
+                    if (remoteClient.sendCommandLineArguments(args)) {
+                        // So we assume it's all taken care of, and quit.
+                        // Output to both to the log and the screen. Therefore, we do not have an additional System.out.println.
+                        LOGGER.info("Arguments passed on to running JabRef instance. Shutting down.");
+                    } else {
+                        LOGGER.warn("Could not communicate with other running JabRef instance.");
+                    }
                 }
+                // We do not launch a new instance in presence of an error
```
Comment (koppor) [inline src/main/java/org/jabref/Launcher.java:192]: This is not an error.

```suggestion
                // We do not launch a new instance if there is another instance running
```

**Review by koppor — CHANGES_REQUESTED** [review]
Thank you for the follow-ups.

I found another thing which is strange ^^

File: src/main/java/org/jabref/Launcher.java (hunk 5 above)
Comment (leaf-soba) [reply src/main/java/org/jabref/Launcher.java]: This comment was not written by me, but I think your comment is better than original one, so I changed it anyway.

**Review by koppor — APPROVED** [review]

**Pull request merged**